File: backtracking/combination_sum.py
def combinationSum(candidates: list[int], target: int) -> list[list[int]]:
    def backtrack(path: list[int], start: int, curr: int):
        if curr == target:
            ans.append(path[:])
            return

        for i in range(start, len(candidates)):
            num = candidates[i]
            if curr + num <= target:
                path.append(num)
                backtrack(path, i, curr + num)
                path.pop()

    ans = []
    backtrack([], 0, 0)
    return ans

File: backtracking/combination_sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @returns {number[][]}
 */
var combinationSum = function(candidates, target) {
    const backtrack = (path, start, curr) => {
        if (curr == target)
        {
            ans.push([...path]);
            return;
        }

        for (let i = start; i < candidates.length; i++)
        {
            const num = candidates[i];
            if (curr + num <= target)
            {
                path.push(num);
                backtrack(path, i, curr + num);
                path.pop();
            }
        }
    }

    let ans = [];
    backtrack([], 0, 0);
    return ans;
};
